File: packages/python/taiwan_payroll/engine/withholding.py
from __future__ import annotations

from .._rounding import apply_rate, parse_rate, round_half_up, not_finite_number
from .._types import WithholdingInput, WithholdingResult


def _assert_non_neg(name: str, v: float) -> None:
    if not_finite_number(v) or v < 0:
        raise ValueError(f"{name} must be a finite non-negative number, got {v}")


def calc_withholding(data: dict, inp: WithholdingInput) -> WithholdingResult:
    it = data.get("incomeTax")
    if not it:
        raise ValueError(f"No incomeTax data for year {data.get('year')}（該年度無所得稅扣繳資料）")

    if inp.type == "resident":
        _assert_non_neg("monthlySalary", inp.monthly_salary)
        dep = inp.dependents
        _assert_non_neg("dependents", dep)
        deductions = it["residentExemption"] * (1 + dep) + it["standardDeduction"] + it["salaryDeduction"]
        taxable = int(max(0, inp.monthly_salary * 12 - deductions))
        bracket = next((b for b in it["brackets"] if taxable >= b["min"] and (b["max"] is None or taxable <= b["max"])), None)
        if bracket is None:
            raise ValueError(f"No tax bracket for taxable {taxable}")
        rn, rd = parse_rate(bracket["rate"])
        annual_tax = max(0, round_half_up(taxable * rn - bracket["progressiveDiff"] * rd, rd))
        return WithholdingResult(withholding=round_half_up(annual_tax, 12), rate=bracket["rate"], taxable_annual=taxable)

    if inp.type == "residentBonus":
        _assert_non_neg("amount", inp.amount)
        rate = it["nonMonthly"]["rate"]
        thr = it["nonMonthly"]["threshold"]
        wh = apply_rate(int(inp.amount), [rate], "round") if inp.amount >= thr else 0
        return WithholdingResult(withholding=wh, rate=rate)

    # nonResident
    _assert_non_neg("monthlySalary", inp.monthly_salary)
    threshold = it["nonResident"]["reducedThresholdMultiplier"] * data["minimumWage"]["monthly"]
    rate = it["nonResident"]["reducedRate"] if inp.monthly_salary <= threshold else it["nonResident"]["rate"]
    return WithholdingResult(withholding=apply_rate(int(inp.monthly_salary), [rate], "round"), rate=rate)

File: packages/core/src/engine/withholding.ts
import type { YearData, WithholdingInput, WithholdingResult } from '../types';
import { applyRate, parseRate, roundHalfUp } from '../rounding/strategies';

function assertNonNeg(name: string, v: number): void {
  if (!Number.isFinite(v) || v < 0) {
    throw new Error(`${name} must be a finite non-negative number, got ${v}`);
  }
}

/**
 * 薪資所得扣繳稅額。三條路徑：
 *  - resident：公式法。應稅=max(0, 月薪×12 − (免稅額×(1+扶養)+標準扣除+薪資扣除))；
 *    全年稅=max(0, round(應稅×率−累進差額))；每月=round(全年稅/12)。兩步皆 round-half-up。
 *  - residentBonus：非每月給付，達起扣標準按給付額 5%，否則 0。
 *  - nonResident：月薪 ≤ 1.5×基本工資 → 6%，否則 18%。
 * Source: 財政部 115 年度薪資所得扣繳稅額表說明／各類所得扣繳率標準。
 */
export function calcWithholding(data: YearData, input: WithholdingInput): WithholdingResult {
  const it = data.incomeTax;
  if (!it) throw new Error(`No incomeTax data for year ${data.year}（該年度無所得稅扣繳資料）`);

  if (input.type === 'resident') {
    assertNonNeg('monthlySalary', input.monthlySalary);
    const dependents = input.dependents ?? 0;
    assertNonNeg('dependents', dependents);
    const deductions = it.residentExemption * (1 + dependents) + it.standardDeduction + it.salaryDeduction;
    // 應稅所得以整數 NTD 計（與 Python 端 int() 契約一致）。
    const taxable = Math.floor(Math.max(0, input.monthlySalary * 12 - deductions));
    const bracket = it.brackets.find((b) => taxable >= b.min && (b.max === null || taxable <= b.max));
    if (!bracket) throw new Error(`No tax bracket for taxable ${taxable}`);
    const { num: rNum, den: rDen } = parseRate(bracket.rate);
    const annualTax = Math.max(0, roundHalfUp(taxable * rNum - bracket.progressiveDiff * rDen, rDen));
    return { withholding: roundHalfUp(annualTax, 12), rate: bracket.rate, taxableAnnual: taxable };
  }

  if (input.type === 'residentBonus') {
    assertNonNeg('amount', input.amount);
    const { rate, threshold } = it.nonMonthly;
    return { withholding: input.amount >= threshold ? applyRate(Math.floor(input.amount), [rate], 'round') : 0, rate };
  }

  // nonResident
  assertNonNeg('monthlySalary', input.monthlySalary);
  const threshold = it.nonResident.reducedThresholdMultiplier * data.minimumWage.monthly;
  const rate = input.monthlySalary <= threshold ? it.nonResident.reducedRate : it.nonResident.rate;
  return { withholding: applyRate(Math.floor(input.monthlySalary), [rate], 'round'), rate };
}
